### ISSUE TITLE
cloneNode vs createNode

### ISSUE THREAD

**Issue by jamesplease** [ChrisAntaki/dom-pool#4]
The `create` method always makes a new node. Cloning from a template node should be faster for most elements.

One idea to implement this would be creating a template when the Pool is created using the tagName. Then `create` would clone the template.

Something like:

``` diff
function Pool(params) {
    ...
    this.tagName = params.tagName.toLowerCase();
+   this.template = this.createTemplate();  
    this.namespace = params.namespace;
}

+ Pool.prototype.createTemplate = function() {
+  if (this.namespace) {
+       return document.createElementNS(this.namespace, this.tagName);
+   } else {
+       return document.createElement(this.tagName);
+    }
+ }

Pool.prototype.create = function() {
-   if (this.namespace) {
-       return document.createElementNS(this.namespace, this.tagName);
-   } else {
-       return document.createElement(this.tagName);
-   }
+   return this.template.cloneNode(false);
};
```

I can make a PR if you're interested.

**Comment by ChrisAntaki** [comment]
Awesome! Would you have time to create a JSPerf benchmark of this?

**Comment by jamesplease** [comment]
Sure. I'll do that now.

**Comment by ChrisAntaki** [comment]
Thanks!

**Comment by jamesplease** [comment]
Alright, JSPerf's charts are super annoying to read, but [here it is](http://jsperf.com/clone-vs-create-for-pool).

The tl;dr is that cloning nodes in Safari and Firefox is always either comparable to, or faster than, creating them. So for those browsers it's preferable to clone.

With Chrome, it's not so black and white: it's comparable/faster as certain nodes (divs and svg's), but slower at other nodes, like imgs and inputs.

This is consistent with what I remember about Chrome's behavior. Since I rarely clone "special" nodes like inputs or img's, I tend to always use `cloneNode`.

This may be going overboard, but there could be a flag set in the constructor.

``` js
var pool = new Pool({
  tagName: 'img',
  useClone: false
});
```

Let me know if you're interested in updating the library based on these results, and I'll be happy to put together a PR!

**Comment by ChrisAntaki** [comment]
Great, I'll do some more benchmarks with mobile devices soon. It appears JSPerf's servers might currently be on fire. Hopefully later today they'll have it sorted out.

**Comment by jamesplease** [comment]
Sounds like a plan! :v:

**Comment by ChrisAntaki** [comment]
Interestingly, Android's Chrome mobile seems to create `div`s faster than it clones them. iOS' Safari mobile is flipped.

Would you be down to make a PR that had the cloning feature turned off by default, but had a configuration switch like "generation" === "clone" || "create" during instantiation? This would allow for someone to turn on the feature at the best times, like serving a webpage to Safari, compiling a Cordova app for iOS, etc.

**Comment by ChrisAntaki** [comment]
Actually, the `useClone` Boolean parameter is probably more efficient

**Comment by jamesplease** [comment]
Sure thing! If I don't get to it tonight, I'll get to it in a day or two.

**Comment by ChrisAntaki** [comment]
Thanks man
